### PR TITLE
new example: Burgers equation solution with MPDATA

### PR DESCRIPTION
**Pull request by WojciechNeuman**

### CODE REVIEW

**Comment by Tosiekdev** [comment]
do precommit install

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            u[i] = -np.sin(np.pi * x0)
+        else:
+            if xi == 0:
+                u[i] = 0
+            else:
+                x0 = find_root(x0=xi/abs(xi), t=t, xi=xi)
+                u[i] = -np.sin(np.pi * x0)
+    return u
+
+def initialize_simulation(nt, nx, t_max):
```
Comment (MyNameIsArko) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:13]: Burger equation is often demonstrated with periodic boundary conditions, which better reflect the physics for wave propagation and conservation. Maybe adding and testing periodic boundary conditions would enhance the example?

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    """
+    The derivative of the function f with respect to x0.
+    """
+    return 1 - np.cos(np.pi*x0)*np.pi*t
+
+def find_root(x0, t, xi):
+    """Find the root of the equation f(x0, t, xi) = 0 using Newton's method."""
+    return root_scalar(f, args=(t, xi), x0=x0, method='newton', maxiter=1000, fprime=df).root
+
+def analytical_solution(x, t):
```
Comment (MyNameIsArko) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: Transition between pre-shock and post-shock regimes using T_SHOCK is abrupt and the code may not handle multi-valued post-shock solutions or entropy conditions properly. Maybe add comments clarifying the assumptions behind the analytical solution, especially for the post-shock regime. Also consider warning users if the simulation time exceeds the shock formation time (`T_SHOCK`), as the analytical solution may no longer be unique or physically accurate.

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py
Comment (MyNameIsArko) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: There is code duplication for functions `analytical_solution`, `initialize_simulation`, `update_advector_n`, `calculate_analytical_solutions`, `run_numerical_simulation`, `plot_analytical_solutions`, `plot_numerical_vs_analytical` and `plot_gif`. Better would be to extract them into some common file or just import them in notebook from .py file

File: tests/unit_tests/test_burgers_equation.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import numpy as np
+from examples.PyMPDATA_examples.burgers_equation.burgers_equation import run_numerical_simulation
+
+
+class TestGeneralNumericalVeryfication(unittest.TestCase):
+    """Test suite for general numerical verification of Burgers' equation simulation."""
+
+    def setUp(self):
+        """Run the simulation once for all tests."""
+        self.states, self.x, self.dt, self.dx = run_numerical_simulation(nt=400, nx=100, t_max=1/np.pi)
```
Comment (WiktorProsowicz) [inline tests/unit_tests/test_burgers_equation.py]: Running the simulation core in the test's setup function entails the possibility of raising an unhandled error, which is going to ruin all tests without an informative feedback. Consider putting here some exception-handling statements so that the test informs the user that there was a failure in the setup and what is it related to.

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    )
+    advector = VectorField(
+        data=(np.full(courants_x.shape, 0.0),),
+        halo=OPTIONS.n_halo,
+        boundary_conditions=(Constant(0), Constant(0))
+    )
+    solver = Solver(stepper=stepper, advectee=advectee, advector=advector)
+    return dt, dx, x, advectee, advector, solver
+
+def update_advector_n(vel, dt, dx, slice_idx):
```
Comment (yancostrishevsky) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:35]: In the `update_advector_n` function, you're computing updated advector values at each simulation step. Have you verified how close these computed Courant numbers are to the stability threshold (Courant ≤ 1), particularly under extreme conditions such as near the shock time?

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    """
+    return x0 - np.sin(np.pi*x0)*t - xi
+
+def df(x0, t, _):
+    """
+    The derivative of the function f with respect to x0.
+    """
+    return 1 - np.cos(np.pi*x0)*np.pi*t
+
+def find_root(x0, t, xi):
```
Comment (yancostrishevsky) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: You've introduced Newton's method in `find_root` to determine analytical solution roots. Have you tested its convergence robustness in scenarios where gradients are sharp (particularly around shock formation) or when the initial guess x0 is far from the actual root or near points of discontinuity?

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+class One:
+    """A class to represent a slice that can be added or subtracted from"""
+
+    def __radd__(self, other):
+        """"""
+        return slice(other.start + 1, other.stop + 1)
+
+    def __rsub__(self, other):
+        """"""
+        return slice(other.start - 1, other.stop - 1)
```
Comment (ogioldat) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: I couldn't find any place you use this class, do we need it?

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py
Comment (ogioldat) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: You can consider type annotations for functions — both for params and returned value.

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+def find_root(x0, t, xi):
+    """Find the root of the equation f(x0, t, xi) = 0 using Newton's method."""
+    return root_scalar(f, args=(t, xi), x0=x0, method='newton', maxiter=1000, fprime=df).root
+
+def analytical_solution(x, t):
+    """
+    Analytical solution for the wave equation
+    """
+    u = np.zeros(len(x))
+    for i, xi in enumerate(x):
```
Comment (WiktorProsowicz) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: In the analytical_solution function, the find_root function is called for each element in x, yet each consecutive call of the function does not depend on the previous one, i.e. they can be run independently. Using a loop here can be computationally expensive for large arrays. Consider parallelizing this loop or using compiled operations if possible.

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        if t < T_SHOCK:
+            x0 = find_root(x0=0, t=t, xi=xi)
+            u[i] = -np.sin(np.pi * x0)
+        else:
+            if xi == 0:
+                u[i] = 0
+            else:
+                x0 = find_root(x0=xi/abs(xi), t=t, xi=xi)
```
Comment (ogioldat) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: The initial guess for `x0` here is `xi/abs(xi)`, which will be either `1` or `-1`. A short comment explaining the rationale behind this initial guess and why it's suitable for the shock's behavior would be beneficial.

**Comment by OguzhanBerkeOzdil** [comment]
I didn't see any minimal README for guidance so maybe something you might think to add.

[README.md](https://github.com/user-attachments/files/20774692/README.md)

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py
Comment (OguzhanBerkeOzdil) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: Docstring mismatch in ``def plot_analytical_solutions(solutions, t_range, figsize=(5, 5)):`` you can change as a ``def plot_analytical_solutions(solutions, t_range, filename="analytical", figsize=(5, 5)):`` it will fix the ``plot_analytical_solutions`` signature and docstring to mention filename and the internal use of ``X_ANAL``

Comment (OguzhanBerkeOzdil) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: Docstring mismatch in ``def plot_analytical_solutions(solutions, t_range, figsize=(5, 5)):`` you can change as a ``def plot_analytical_solutions(solutions, t_range, filename="analytical", figsize=(5, 5)):`` it will fix the ``plot_analytical_solutions`` signature and docstring to mention filename and the internal use of ``X_ANAL``

Comment (OguzhanBerkeOzdil) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: you can a conditional import, so tests skip if required modules are unavailable and updated the import path to load from PyMPDATA_examples.burgers_equation.burgers_equation

try:
    import numpy as np
    from PyMPDATA_examples.burgers_equation.burgers_equation import run_numerical_simulation
    MODULES_AVAILABLE = True
except ModuleNotFoundError:
    MODULES_AVAILABLE = False


@unittest.skipUnless(MODULES_AVAILABLE, "Required modules not installed")

Comment (OguzhanBerkeOzdil) [inline examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: maybe might think to fix tests and spelling mistakes

``Veryfication`` → ``Verification`` and ``analitical`` → ``analytical``.

File: tests/unit_tests/test_burgers_equation.py
Comment (OguzhanBerkeOzdil) [inline tests/unit_tests/test_burgers_equation.py:1]: The docstring calls this a “periodic boundary,” but the code applies Constant(0), that’s homogeneous Dirichlet (u = 0 at both ends), not periodic.

Consider to update the description to “homogeneous Dirichlet” (and assert the end values stay zero) or implement a genuine periodic BC.

File: tests/unit_tests/test_burgers_equation.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# pylint: disable=line-too-long
```
Comment (ogioldat) [inline tests/unit_tests/test_burgers_equation.py]: Any good reason for disabling lint rule?

**Comment by slayoo** [comment]
pylint complains a bit here (https://github.com/open-atmos/PyMPDATA/actions/runs/15541208707/job/43825449294?pr=577):
```
************* Module tests.unit_tests.test_burgers_equation
tests/unit_tests/test_burgers_equation.py:14:21: C0103: Attribute name "x" doesn't conform to snake_case naming style (invalid-name)
tests/unit_tests/test_burgers_equation.py:14:29: C0103: Attribute name "dt" doesn't conform to snake_case naming style (invalid-name)
tests/unit_tests/test_burgers_equation.py:14:38: C0103: Attribute name "dx" doesn't conform to snake_case naming style (invalid-name)

-----------------------------------
Your code has been rated at 9.99/10
```

to make `pre-commit` do its job, please either set it up for automatic runs with:
```
pip install pre-commit
cd PyMPDATA # the cloned repo
pre-commit install
```
or trigger it manually before committing with:
```
pre-commit run --all-files
```

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py (hunk 6 above)
Comment (WojciechNeuman) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: You are right. At the beginning we were in fact using this class, but during refactoring we stopped using it. I removed this obsolete class

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py
Comment (WojciechNeuman) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: Thank you for the suggestion. I've fixed the spelling mistakes

Comment (PPojda) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:1]: changed here: https://github.com/open-atmos/PyMPDATA/pull/577/commits/eeba715ed2b2bf54acd4e8375ab6d378f59aa0c1

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py (hunk 1 above)
Comment (Tosiekdev) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:13]: Unfortunately we couldn't found any analytical solution to Burgers' equation which has periodic boundaries and is not too complicated mathematically. Moreover, after creation of the shock, velocity conservation does not hold, and PyMPDATA is conservative solver, so there is no use to show scenario with a wave propagation.

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py (hunk 8 above)
Comment (michalszc) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: Solved in: [7f473e9](https://github.com/open-atmos/PyMPDATA/pull/577/commits/7f473e90917643ab8b74e28234b830e0649c07e2)

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py (hunk 2 above)
Comment (Tosiekdev) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: This analytical solution is coded for very specific case covered in our presentations during classes. It is proper solution, and the only problem which may occur in this solution are numerical imperfections implied by the nature of floating point arithmetic.  But I was tempted  to add a comment about initial and boundary conditions of our analytical solution: [96e5837](https://github.com/open-atmos/PyMPDATA/pull/577/commits/96e583734d36a49403d421067f088910d43f4926)

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py (hunk 4 above)
Comment (Iriide) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py:35]: We have an assert outside of this function checking the computed Courant number. We think it should be sufficient.
```python
assert np.all(advector.get_component(0) <= 1)
```

File: examples/PyMPDATA_examples/burgers_equation/burgers_equation.py (hunk 5 above)
Comment (Iriide) [reply examples/PyMPDATA_examples/burgers_equation/burgers_equation.py]: We have tested this functions in various settings. The main factor contributing to stability is the ```maxiter``` argument, which is currently set to 1000. It was decided as a best compromise between computation accuracy and performance.

**Comment by slayoo** [comment]
queued forever... let's try to close & reopen

**Review by slayoo — DISMISSED** [review]
